### PR TITLE
(For discussion) - Move authorization code to Authorizer, add authorized_only API

### PR DESCRIPTION
**Pull request by kevinrobinson**
This is for discussion with https://github.com/studentinsights/studentinsights/issues/1080.  This has two commits.

The first commit adds an Authorizer class, with a minimal implementation of what was called the "blocks-based API" in https://github.com/studentinsights/studentinsights/issues/1080.  The comments and tests there are consistent, and sketch the idea and usage.

The second commit is aimed at moving authorization code into the Authorizer class.  It doesn't update the call sites, and so Educator methods become shims into the Authorizer method calls for now.  This reveals in the Authorizer methods how there's different implementations of the authorization rules (even just from gathering together the methods in Educator, without looking at controller code yet).

The intention here is to give folks something concrete to look at and give feedback on in regards to https://github.com/studentinsights/studentinsights/issues/1080.  If this seems like a good direction, the next step might be "How would this work for the students_controller?" or "How would this affect performance?"  Those might be two additional commits, and then a natural point to decide about merging a chunk of work, and listing out what else would be needed to finish this off altogether.

### CODE REVIEW

File: app/models/educator.rb [hunk 1]
```diff
@@ -128,39 +70,17 @@ def default_section
...
-      Section.all
-    elsif schoolwide_access?
-      Section.joins(:course).where('courses.school_id = ?', school.id)
-    else
-      sections
-    end
-  end
-
-  def allowed_homerooms_by_name
-    allowed_homerooms.order(:name)
```
Comment (kevinrobinson) [inline app/models/educator.rb]: Removing this.

File: app/models/educator.rb [hunk 2]
```diff
@@ -17,13 +17,14 @@ class Educator < ActiveRecord::Base
 
   validates :email, presence: true, uniqueness: true
 
-  validate :has_school_unless_districtwide,
-           :admin_gets_access_to_all_students,
-           :grade_level_access_is_array_of_strings,
-           :grade_level_strings_are_valid,
-           :grade_level_strings_are_uniq,
-           :grade_level_access_is_not_nil
```
Comment (kevinrobinson) [inline app/models/educator.rb]: These look like they're all for validation (some read like authorization), so I moved them to private and prefixed them all with `validate`.

File: app/lib/authorizer.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    return true if @educator.districtwide_access?
+
+    return false if @educator.school.present? && @educator.school != section.course.school
+
+    return true if @educator.schoolwide_access? || @educator.admin?
+    return true if section.in?(@educator.sections)
+    false
+  end
+
+  # TODO(kr) remove implementation
```
Comment (kevinrobinson) [inline app/lib/authorizer.rb]: These intention for these methods is that we'd be able to remove them, and defined them in terms of the other methods above.  That for this one in particular we'd say:

```
students = authorized_only { School.find(params[:id]).students }
```

in the controller.

File: spec/lib/authorizer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  end
+
+  it 'test context is setup correctly' do
+    expect(School.all.size).to eq 12
+    expect(Homeroom.all.size).to eq 2
+    expect(Student.unscoped.all.size).to eq 2
+    expect(Educator.all.size).to eq 3
+  end
+
+  describe 'Student' do
```
Comment (kevinrobinson) [inline spec/lib/authorizer_spec.rb]: This shows intended usage of `authorized_only`.

File: spec/lib/authorizer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    end
+
+    it 'limits access for arrays' do
+      expect(authorized_only(healey_teacher) { [shs_student, healey_kindergarten_student] }).to eq [healey_kindergarten_student]
+    end
+  end
+
+  describe 'other models' do
+    it 'raises on calls with unchecked models' do
+      expect { authorized_only(districtwide_admin) { Educator.all } }.to raise_error(Exceptions::EducatorNotAuthorized)
```
Comment (kevinrobinson) [inline spec/lib/authorizer_spec.rb]: A catch for if someone passes a model or query that isn't supported by the authorization layer.

**Comment by alexsoble** [comment]
Hey @kevinrobinson, this is awesome. I love the way the `Authorizer` class pulls together all the authorization-logic related methods like `is_authorized_for_student?`, `is_authorized_for_school?`, `is_authorized_for_section?`. 

It feels useful, centralized, and very object-oriented. One specific use case: we should add `authorizer.rb` to the notification bot, or whatever tool we may use to look for PRs that touch auth and need careful review. And now new contributors should never have to write new auth code from scratch! We'll want to make "how to use the authorization API" an important part of onboarding. 

I'll add a few inline code comments where I have questions. Overall I am super supportive of and excited about this pull request! I think this is one of our highest priorities right now.

I was thinking about how folks could best support this PR, maybe one way is by helping test/QA the performance changes this may introduce so we can make sure it doesn't introduce any large regressions.

File: app/models/educator.rb (hunk 2 above)
Comment (alexsoble) [reply app/models/educator.rb]: Yes, these are all for validation. Question here: does prefixing with `validation` automatically run these as validations? Is this some ancient Rails magic you're invoking here? 🔮

File: app/lib/authorized_only.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  # and by the Model class.
+  def dispatch(&block)
+    return_value = block.call
+
+    # Can't duck type here, since other things like hashes
+    # are iterable or mappable in ways we don't want to
+    # support.
+    if return_value.class < ActiveRecord::Relation # subclass?
+      filter_array(return_value.to_a)
+    elsif return_value.class == Array # subclass?
```
Comment (alexsoble) [inline app/lib/authorized_only.rb]: These look swapped to me. It calls `filter_relation` for array and `filter_array` for relation.

File: app/lib/authorizer.rb [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+# Standalone usage:
+#   authorizer = Authorizer.new(educator)
+#   student = authorizer.authorized_only { Student.find(params[:id]) }
+#   homeroom = authorizer.authorized_only { Homeroom.(params[:id]) }
+#   students = authorizer.authorized_only { homeroom.students } 
+#
+# Instead of:
+#   Homeroom.find(id).students # no authorization check
+#   Section.find(id).students # no authorization check
+#   current_educator.homeroom.students # doesn't consider other roles
```
Comment (alexsoble) [inline app/lib/authorizer.rb]: @kevinrobinson What do you mean by "doesn't consider other roles"?

File: app/lib/authorized_only.rb (hunk 6 above)
Comment (kevinrobinson) [reply app/lib/authorized_only.rb]: thanks!

File: app/lib/authorizer.rb (hunk 7 above)
Comment (kevinrobinson) [reply app/lib/authorizer.rb]: In other words, for a given homeroom page, this doesn't do what you'd want for all user types (eg, district admin).

File: app/models/educator.rb (hunk 2 above)
Comment (kevinrobinson) [reply app/models/educator.rb]: No I just renamed them :)

**Comment by kevinrobinson** [comment]
@alexsoble I'm traveling this week so won't be able to do much more here until next week at the earliest.  In terms of a todolist I think it's:
[ ] Does this seem like a good path?
[ ] How would this work for the students_controller?
[ ] How would this affect performance?
[ ] Pause, reflect, merge a chunk of work, make the next list of todos

File: app/lib/authorizer.rb (hunk 7 above)
Comment (alexsoble) [reply app/lib/authorizer.rb]: Ooooooh right, that makes sense. For example, if `current_educator` is a district admin s/he won't be assigned a `homeroom`.

**Comment by alexsoble** [comment]
@kevinrobinson Thoughts a couple of your TODO items: 

**Does this seem like a good path?** 

Yes!

**How would this work for the students_controller?**

My first reaction is that we would call something like `student = authorizer.authorized_only { Student.find(params[:id]) }`, but I think you mean something deeper than "mechanically, what code will we write to implement this in the students controller?" Could you say a little more about what you mean?

**Comment by kevinrobinson** [comment]
Pulled off part of this work into: https://github.com/studentinsights/studentinsights/pull/1274.

@alexsoble re: controller code, I started doing this on the `students_controller` as a spike, you can check it out [here](https://github.com/studentinsights/studentinsights/pull/1237/files#diff-11751d644b56a1cce6d8dfc73e331b91L1).  There's some messiness that came up (event_note authorization is nested and works differently, partial models don't work with the authorization method), so I cut all that out for now in favor of making progress on https://github.com/studentinsights/studentinsights/pull/1274, but if you have any feedback let me know!

**Comment by kevinrobinson** [comment]
There is additional spike code on this branch https://github.com/studentinsights/studentinsights/compare/refactor/authorization, using the work in https://github.com/studentinsights/studentinsights/pull/1274 to rework the `students_controller`.  That ended up being too large of a change so I split out https://github.com/studentinsights/studentinsights/pull/1274, but that's there for looking at for future work in this direction.

**Pull request closed without merging**